<a href="https://colab.research.google.com/github/amiguelnobrega/Waze/blob/main/waze1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#import pandas and numpy libraries with aliases
import pandas as pd
import numpy as np

In [ ]:
#mount google drive in colab notebook
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [ ]:
df = pd.read_csv('/content/gdrive/Othercomputers/My MacBook Air/Estudo/Google Advance DA/Projects/Waze/waze_dataset.csv')
print('done!')

done!


In [ ]:
df.head(10)

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   
5   5  retained       113     103      279.544437                     2637   
6   6  retained         3       2      236.725314                      360   
7   7  retained        39      35      176.072845                     2999   
8   8  retained        57      46      183.532018                      424   
9   9   churned        84      68      244.802115                     2997   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   
5                       0                       0        901.238699   
6                     185                      18       5249.172828   
7                       0                       0       7892.052468   
8                       0                      26       2651.709764   
9                      72                       0       6043.460295   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android  
5               439.101397             15            11   iPhone  
6               726.577205             28            23   iPhone  
7              2466.981741             22            20   iPhone  
8              1594.342984             25            20  Android  
9              2341.838528              7             3   iPhone

In [ ]:
df.tail(10)

ID     label  sessions  drives  total_sessions  \
14989  14989  retained         6       6      153.373616   
14990  14990   churned        73      61      329.904300   
14991  14991   churned        50      41      102.444592   
14992  14992  retained       112      90      267.040136   
14993  14993       NaN        67      57       97.570074   
14994  14994  retained        60      55      207.875622   
14995  14995  retained        42      35      187.670313   
14996  14996  retained       273     219      422.017241   
14997  14997   churned       149     120      180.524184   
14998  14998  retained        73      58      353.419797   

       n_days_after_onboarding  total_navigations_fav1  \
14989                     2988                     371   
14990                      614                      60   
14991                     1463                       0   
14992                     1283                      14   
14993                     1131                     207   
14994                      140                     317   
14995                     2505                      15   
14996                     1873                      17   
14997                     3150                      45   
14998                     3383                      13   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
14989                      24       1553.284152               833.421155   
14990                      46       6090.450154              3323.880771   
14991                     121       4094.536313              2201.984210   
14992                       0       1525.932143              1116.650478   
14993                     102       2267.052913               318.120634   
14994                       0       2890.496901              2186.155708   
14995                      10       4062.575194              1208.583193   
14996                       0       3097.825028              1031.278706   
14997                       0       4051.758549               254.187763   
14998                      51       6030.498773              3042.436423   

       activity_days  driving_days   device  
14989             23            20   iPhone  
14990              0             0  Android  
14991              4             2   iPhone  
14992              5             5   iPhone  
14993             27            26   iPhone  
14994             25            17   iPhone  
14995             25            20  Android  
14996             18            17   iPhone  
14997              6             6   iPhone  
14998             14            13   iPhone

In [ ]:
df.info()

#some values are missing from the label column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


In [ ]:
df.describe()

ID      sessions        drives  total_sessions  \
count  14999.000000  14999.000000  14999.000000    14999.000000   
mean    7499.000000     80.633776     67.281152      189.964447   
std     4329.982679     80.699065     65.913872      136.405128   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.661156   
50%     7499.000000     56.000000     48.000000      159.568115   
75%    11248.500000    112.000000     93.000000      254.192341   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14999.000000            14999.000000   
mean               1749.837789              121.605974   
std                1008.513876              148.121544   
min                   4.000000                0.000000   
25%                 878.000000                9.000000   
50%                1741.000000               71.000000   
75%                2623.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14999.000000      14999.000000             14999.000000   
mean                29.672512       4039.340921              1860.976012   
std                 45.394651       2502.149334              1446.702288   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2212.600607               835.996260   
50%                  9.000000       3493.858085              1478.249859   
75%                 43.000000       5289.861262              2464.362632   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14999.000000  14999.000000  
mean       15.537102     12.179879  
std         9.004655      7.824036  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

In [ ]:
#to know the considered devices (only mobile)
df.device.unique()

array(['Android', 'iPhone'], dtype=object)

In [ ]:
#isolating null values from the label column and retrieve summary stats of the same
null_df = df[df['label'].isnull()]

null_df.describe()

ID    sessions      drives  total_sessions  \
count    700.000000  700.000000  700.000000      700.000000   
mean    7405.584286   80.837143   67.798571      198.483348   
std     4306.900234   79.987440   65.271926      140.561715   
min       77.000000    0.000000    0.000000        5.582648   
25%     3744.500000   23.000000   20.000000       94.056340   
50%     7443.000000   56.000000   47.500000      177.255925   
75%    11007.000000  112.250000   94.000000      266.058022   
max    14993.000000  556.000000  445.000000     1076.879741   

       n_days_after_onboarding  total_navigations_fav1  \
count               700.000000              700.000000   
mean               1709.295714              118.717143   
std                1005.306562              156.308140   
min                  16.000000                0.000000   
25%                 869.000000                4.000000   
50%                1650.500000               62.500000   
75%                2508.750000              169.250000   
max                3498.000000             1096.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count              700.000000        700.000000               700.000000   
mean                30.371429       3935.967029              1795.123358   
std                 46.306984       2443.107121              1419.242246   
min                  0.000000        290.119811                66.588493   
25%                  0.000000       2119.344818               779.009271   
50%                 10.000000       3421.156721              1414.966279   
75%                 43.000000       5166.097373              2443.955404   
max                352.000000      15135.391280              9746.253023   

       activity_days  driving_days  
count     700.000000    700.000000  
mean       15.382857     12.125714  
std         8.772714      7.626373  
min         0.000000      0.000000  
25%         8.000000      6.000000  
50%        15.000000     12.000000  
75%        23.000000     18.000000  
max        31.000000     30.000000

In [ ]:
#isolating non-null values to compare statistics against null values exctract
non_null_df = df[~df['label'].isnull()]

non_null_df.describe()

#no significant differences between the two sets. the mean and standard deviation are fairly consistent

ID      sessions        drives  total_sessions  \
count  14299.000000  14299.000000  14299.000000    14299.000000   
mean    7503.573117     80.623820     67.255822      189.547409   
std     4331.207621     80.736502     65.947295      136.189764   
min        0.000000      0.000000      0.000000        0.220211   
25%     3749.500000     23.000000     20.000000       90.457733   
50%     7504.000000     56.000000     48.000000      158.718571   
75%    11257.500000    111.000000     93.000000      253.540450   
max    14998.000000    743.000000    596.000000     1216.154633   

       n_days_after_onboarding  total_navigations_fav1  \
count             14299.000000            14299.000000   
mean               1751.822505              121.747395   
std                1008.663834              147.713428   
min                   4.000000                0.000000   
25%                 878.500000               10.000000   
50%                1749.000000               71.000000   
75%                2627.500000              178.000000   
max                3500.000000             1236.000000   

       total_navigations_fav2  driven_km_drives  duration_minutes_drives  \
count            14299.000000      14299.000000             14299.000000   
mean                29.638296       4044.401535              1864.199794   
std                 45.350890       2504.977970              1448.005047   
min                  0.000000         60.441250                18.282082   
25%                  0.000000       2217.319909               840.181344   
50%                  9.000000       3496.545617              1479.394387   
75%                 43.000000       5299.972162              2466.928876   
max                415.000000      21183.401890             15851.727160   

       activity_days  driving_days  
count   14299.000000  14299.000000  
mean       15.544653     12.182530  
std         9.016088      7.833835  
min         0.000000      0.000000  
25%         8.000000      5.000000  
50%        16.000000     12.000000  
75%        23.000000     19.000000  
max        31.000000     30.000000

In [ ]:
#null values breakdown per device
null_df['device'].value_counts()

device
iPhone     447
Android    253
Name: count, dtype: int64

In [ ]:
#percentage of null values per each device
round(null_df['device'].value_counts(normalize=True)*100, ndigits = 2).astype(str) + '%'

device
iPhone     63.86%
Android    36.14%
Name: proportion, dtype: object

In [ ]:
#percentage of device in the dataset
print(round(df['device'].value_counts(normalize=True)*100, ndigits = 2).astype(str) + '%')

#results are similar compared with the null values

device
iPhone     64.48%
Android    35.52%
Name: proportion, dtype: object


In [ ]:
#calculate churned vs retained users count and percentage
print(df['label'].value_counts())
print()
print(round(df['label'].value_counts(normalize=True)*100, ndigits = 2).astype(str) + '%')

label
retained    11763
churned      2536
Name: count, dtype: int64

label
retained    82.26%
churned     17.74%
Name: proportion, dtype: object


In [ ]:
#median of all columns by label (retained vs churned)
df.groupby('label').median(numeric_only=True)

ID  sessions  drives  total_sessions  n_days_after_onboarding  \
label                                                                         
churned   7477.5      59.0    50.0      164.339042                   1321.0   
retained  7509.0      56.0    47.0      157.586756                   1843.0   

          total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
label                                                                        
churned                     84.5                    11.0       3652.655666   
retained                    68.0                     9.0       3464.684614   

          duration_minutes_drives  activity_days  driving_days  
label                                                           
churned               1607.183785            8.0           6.0  
retained              1458.046141           17.0          14.0

Users who churned averaged ~3 more drives in the last month than retained users, but retained users used the app on over twice as many days as churned users in the same time period.

The median churned user drove ~200 more kilometers and 2.5 more hours during the last month than the median retained user.

It seems that churned users had more drives in fewer days, and their trips were farther and longer in duration. Perhaps this is suggestive of a user profile.

In [ ]:
#Calculate the median kilometers per drive in the last month for both retained and churned users

#Add column to df called km_per_drive
df['km_per_drive'] = df['driven_km_drives'] / df['drives']

#Group by 'label', calculate median, and isolate for km_per_drive
median_km_per_drive = df.groupby('label').median(numeric_only=True)[['km_per_drive']]
round(median_km_per_drive, ndigits = 2)

km_per_drive
label                 
churned          74.11
retained         75.01

In [ ]:
#Calculate the median kilometers per driving day in the last month for both retained and churned users

#Add column to df called km_per_drive
df['km_per_driving_days'] = df['driven_km_drives'] / df['driving_days']

#Group by 'label', calculate median, and isolate for km_per_driving_days
median_km_per_driving_days = df.groupby('label').median(numeric_only=True)[['km_per_driving_days']]
round(median_km_per_driving_days, ndigits = 2)

km_per_driving_days
label                        
churned                697.54
retained               289.55

In [ ]:
#median number of drives per driving day for each group

#add column to df drives per driving day
df['drives_per_driving_day'] = df['drives'] / df['driving_days']

#Group by `label`, calculate the median, and isolate for drives per driving day
median_drives_per_driving_day = df.groupby('label').median(numeric_only=True)[['drives_per_driving_day']]
round(median_drives_per_driving_day, ndigits = 2)

drives_per_driving_day
label                           
churned                    10.00
retained                    4.06

The median user who churned drove 698 kilometers each day they drove last month, which is ~240% the per-drive-day distance of retained users. The median churned user had a similarly disproporionate number of drives per drive day compared to retained users.

It is clear from these figures that, regardless of whether a user churned or not, the users represented in this data are serious drivers! It would probably be safe to assume that this data does not represent typical drivers at large. Perhaps the data&mdash;and in particular the sample of churned users&mdash;contains a high proportion of long-haul truckers.

In consideration of how much these users drive, it would be worthwhile to recommend to Waze that they gather more data on these super-drivers. It's possible that the reason for their driving so much is also the reason why the Waze app does not meet their specific set of needs, which may differ from the needs of a more typical driver, such as a commuter.

In [ ]:
#churned and retained users by device type
df.groupby(['label', 'device']).size()

label     device 
churned   Android     891
          iPhone     1645
retained  Android    4183
          iPhone     7580
dtype: int64

In [ ]:
#percentage of churned and retained users by device type
round(df.groupby('label')['device'].value_counts(normalize=True)*100).astype(str) + '%'

label     device 
churned   iPhone     65.0%
          Android    35.0%
retained  iPhone     64.0%
          Android    36.0%
Name: proportion, dtype: object

The ratio of iPhone users and Android users is consistent between the churned group and the retained group, and those ratios are both consistent with the ratio found in the overall dataset.